chains/solana/contracts/programs/ccip-offramp/src/instructions/v1/execute.rs

```diff
@@ -368,104 +418,194 @@ fn internal_execute<'info>(
+}
 
-    let logic_receiver = &remaining_accounts[0];
-    let external_execution_signer = &remaining_accounts[1];
-    let msg_accounts = &remaining_accounts[2..end_index];
+impl<'a> ExecuteReportContextRemainingAccountsLayout<'a> {
+    pub fn should_execute_messaging(&self) -> bool {
+        self.messaging_accounts.is_some()
+    }
 
-    // Validate the derivation of the external_execution_signer and calculate its bump
-    let (expected_signer_key, signer_bump) = Pubkey::find_program_address(
-        &[
-            seed::EXTERNAL_EXECUTION_CONFIG,
-            logic_receiver.key().as_ref(),
-        ],
-        &crate::ID,
-    );
-    require_keys_eq!(
-        external_execution_signer.key(),
-        expected_signer_key,
-        CcipOfframpError::InvalidInputsExternalExecutionSignerAccount
-    );
+    pub fn receiver_and_message_account_keys(&self) -> impl Iterator<Item = Pubkey> {
+        let keys: Vec<Pubkey> = self
+            .messaging_accounts
+            .iter()
+            .flat_map(|accs| {
+                Some(accs.logic_receiver.key())
+                    .into_iter()
+                    .chain(accs.remaining_messaging_accounts.iter().map(|a| a.key()))
+            })
+            .collect();
 
-    // Validate that the bitmap corresponds to the individual writable flags
-    for (i, acc) in msg_accounts.iter().enumerate() {
-        require!(
-            !is_writable(source_bitmap, i as u8) || acc.is_writable,
-            CcipOfframpError::InvalidWritabilityBitmap
-        );
+        keys.into_iter()
+    }
+
+    pub fn get_token_accounts_for<'b>(
+        &'b self,
+        router: Pubkey,
+        fee_quoter: Pubkey,
+        token_receiver: Pubkey,
+        chain_selector: u64,
+        i: usize,
+    ) -> Result<TokenAccounts<'a>> {
+        let accs = self
+            .token_accounts_per_token
+            .get(i)
+            .map(|accounts_per_token| accounts_per_token.accounts)
+            .ok_or(CcipOfframpError::InvalidInputsTokenIndices)?;
+
+        validate_and_parse_token_accounts(
+            token_receiver,
+            chain_selector,
+            router,
+            fee_quoter,
+            Some(crate::ID),
+            accs,
+        )
     }
 
-    Ok(MessagingAccounts {
-        logic_receiver,
-        external_execution_signer,
-        external_execution_signer_bump: signer_bump,
-        remaining_accounts: msg_accounts,
-    })
+    pub fn new(
+        remaining_accounts: &'a [AccountInfo<'a>],
+        token_indices: &[u8],
+        report_is_buffered: bool,
+        source_bitmap: &u64,
+    ) -> Result<Self> {
+        ////////////////////////
+        // Buffering subslice //
+        ////////////////////////
+
+        // First, if the report is buffered, it means the last account is the buffering account.
+        // We can simply remove it from consideration from now on.
+        let (remaining_accounts_without_buffering, buffering_accounts) = if report_is_buffered {
```

Comment:
Wdyt about renaming `buffering_accounts` into `buffering_account`, given that it's at most 1?